### PR TITLE
REF: share _union between DTI/TDI

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -806,6 +806,25 @@ def _fast_union(self, other, sort=None):
         else:
             return left
 
+    def _union(self, other, sort):
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: other can a Union[DTI, TDI]

Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: @simonjayhawkins what's the preferred way of typing as `type(self)`?

Comment (simonjayhawkins) [reply pandas/core/indexes/datetimelike.py]: def _union(self: _T, other: Type[_T], sort):

Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: where does _T come from?  I dont see this pattern anywhere else in the code

Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: i vote for merging and adding annotations in follow-up

Comment (jreback) [reply pandas/core/indexes/datetimelike.py]: followup ok

**Pull request merged**